content/blog/2025-12-16-gvn-final-report.md

```diff
@@ -0,0 +1,1350 @@
+    for i, instr in enumerate(instrs):
+        if i >= first_label_idx:
+            break  # Stop if past entry block
+
+        if instr.get("op") == "id" and "dest" in instr:
+            src = instr.get("args", [])[0]
+            if src in param_names and param_use_count[src] == 1:
+                # Rename parameter to match destination
+                rename_parameter(func, src, instr["dest"])
+                remove_copy(i)
+```
+
+**Critical Safety Check:** The entry block boundary check prevents renaming parameters that are used inside loops. Without this check, the optimizer would incorrectly rename loop-carried copies, breaking programs like `1dconv.bril`.
+
+#### 2.8.3 Constant Copy Fusion
+
+Similarly, constants that were immediately copied also represented unnecessary overhead:
+
+```bril
+i.1: int = const 0;
+res.2: int = id i.1;    ← Unnecessary copy from constant
+```
+
+When a constant is only used once (by a copy instruction), we can fuse them by directly changing the constant's destination to the final SSA name:
+
+```python
+def const_copy_fusion(func: dict):
+    """Fuse const definitions with their single-use copies."""
+    # Find constants used only once
+    use_counts = count_all_uses(func)
+
+    for i, instr in enumerate(instrs):
+        if instr.get("op") == "id":
+            src = instr["args"][0]
+            if is_const(src) and use_counts[src] == 1:
+                # Change const destination to copy destination
+                const_instr["dest"] = instr["dest"]
+                remove_copy(i)
+```
+
+#### 2.8.4 Phi Destination Propagation
+
+A third pattern emerges when we trace the flow from entry block definitions through phi nodes:
+
+```bril
+res.1: int = const 0;
+res.2: int = id res.1;    ← Entry copy for phi
+.loop:
+  res.2: int = phi res.2 from entry, res.3 from body
+```
+
+The entry definition (`res.1`) exists only to serve the phi node. If that definition is only used by the entry copy, we can rename the definition directly to the phi destination name and eliminate the copy:
+
+```python
+def phi_destination_propagation(func: dict):
+    """Propagate phi destination names to entry block definitions."""
+    # For each phi, if the entry argument is defined in the entry block
+    # and only used by the phi, rename the definition to use phi's dest
+    for phi in phi_instructions:
+        entry_arg = phi["args"][entry_index]
+        if defined_in_entry_block(entry_arg) and use_count(entry_arg) == 1:
+            rename_def(entry_arg, phi["dest"])
+```
+
+#### 2.8.5 TDCE Fix for Function Parameters
+
+**The Bug:** After parameter renaming, TDCE would incorrectly remove code that depended on renamed parameters.
+
+**Example:**
+```bril
+@convolve(arr.3: ptr<int>, ...) {  ← Parameter renamed from arr
+  loc.1: ptr<int> = ptradd arr.3 offset;
+  ...
+}
+```
+
+TDCE didn't know `arr.3` was a parameter and removed the `ptradd` instruction.
```

Comment:
Why does this have to do with renaming? Wouldn't this have been a problem with the old names too?